### PR TITLE
[SPARK-21739][SQL]Cast expression should initialize timezoneId when it is called statically to convert something into TimestampType

### PR DESCRIPTION
**Pull request by DonnyZone**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/projects/SPARK/issues/SPARK-21739

This issue is caused by introducing TimeZoneAwareExpression.
When the **Cast** expression converts something into TimestampType, it should be resolved with setting `timezoneId`. In general, it is resolved in LogicalPlan phase.

However, there are still some places that use Cast expression statically to convert datatypes without setting `timezoneId`. In such cases,  `NoSuchElementException: None.get` will be thrown for TimestampType.

This PR is proposed to fix the issue. We have checked the whole project and found two such usages(i.e., in`TableReader` and `HiveTableScanExec`). 

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by DonnyZone** [comment]
@cloud-fan @gatorsmile

**Comment by DonnyZone** [comment]
@cloud-fan @gatorsmile

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
Also cc @ueshin

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala [hunk 1]
```diff
@@ -104,7 +105,7 @@ case class HiveTableScanExec(
     hadoopConf)
 
   private def castFromString(value: String, dataType: DataType) = {
-    Cast(Literal(value), dataType).eval(null)
+    Cast(Literal(value), dataType, Option(SQLConf.get.sessionLocalTimeZone)).eval(null)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Please use `CastSupport.cast`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 2]
```diff
@@ -227,7 +228,8 @@ class HadoopTableReader(
       def fillPartitionKeys(rawPartValues: Array[String], row: InternalRow): Unit = {
         partitionKeyAttrs.foreach { case (attr, ordinal) =>
           val partOrdinal = partitionKeys.indexOf(attr)
-          row(ordinal) = Cast(Literal(rawPartValues(partOrdinal)), attr.dataType).eval(null)
+          row(ordinal) = Cast(Literal(rawPartValues(partOrdinal)), attr.dataType,
+            Option(SQLConf.get.sessionLocalTimeZone)).eval(null)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: The same here.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: We also need a test case for verifying this.

**Comment by gatorsmile** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 2 above)
Comment (DonnyZone) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: OK, I will work on it

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala (hunk 1 above)
Comment (DonnyZone) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: BTW, is it elegant to initialize a `CastSupport` (`DataSourceAnalysis` rule or `DataSourceStrategy`) here, in which we still need to pass `SQLConf`?

Comment (DonnyZone) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveTableScanExec.scala]: Here, we can obtain SQLConf directly with `sparkSession.sessionState.conf`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 2 above)
Comment (DonnyZone) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Do you mean a test case for HadoopTableReader?  a little confusing

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala [hunk 3]
```diff
@@ -68,4 +68,25 @@ class QueryPartitionSuite extends QueryTest with SQLTestUtils with TestHiveSingl
       sql("DROP TABLE IF EXISTS createAndInsertTest")
     }
   }
+
+  test("SPARK-21739: Cast expression should initialize timezoneId " +
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: This test can pass without the change in `TableReader.scala`. We need another test case.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala [hunk 4]
```diff
@@ -68,4 +68,25 @@ class QueryPartitionSuite extends QueryTest with SQLTestUtils with TestHiveSingl
...
+
+    // test for Cast expression in TableReader
+    checkAnswer(sql("select value from table_with_timestamp_partition"),
+      Seq(Row(1), Row(2)))
+
+    // test for Cast expression in HiveTableScanExec
+    checkAnswer(sql("select value from table_with_timestamp_partition " +
+      "where ts = '2010-01-02 00:00:00.000'"), Row(2))
+
+    sql("DROP TABLE IF EXISTS table_with_timestamp_partition")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: use `WithTable`. You can check how we do it in the other test cases

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala (hunk 3 above)
Comment (DonnyZone) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: Oh, it should select the TimestampType column. Thanks for reminder, I will fix it.

**Comment by DonnyZone** [comment]
Test case updated.

**Comment by DonnyZone** [comment]
Moreover, how about using `CastSupport.cast`, shall I initialize a `DataSourceAnalysis` or `DataSourceStrategy` ?

**Comment by gatorsmile** [comment]
Could you check the changes I made before? https://github.com/apache/spark/commit/081d57a82bfd22c244615f4e8f33eaf0427a73df

Thanks!

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2

**Pull request closed without merging**